x/ccv/consumer/keeper/relay.go

```diff
@@ -188,8 +188,11 @@ func (k Keeper) SendPackets(ctx sdk.Context) {
 	pending := k.GetAllPendingPacketsWithIdx(ctx)
 	idxsForDeletion := []uint64{}
 	for _, p := range pending {
+		if !k.PacketSendingPermitted(ctx) {
```

Comment:
Not sure what you mean, `PacketSendingPermitted` could return true for multiple packets in the for-loop (all vsc matured for example). The code is intended to follow [this behavior](https://github.com/cosmos/interchain-security/blob/shawn/throttle-with-retries-consumer-changes/docs/docs/adrs/adr-008-throttle-retries.md#consumer-changes)

Reply:
If the if condition is triggered in the middle of the iteration, then `k.DeletePendingDataPackets(ctx, idxsForDeletion...)` will not be called due to the return statement.

Reply:
Ah great catch! From my understanding,`PacketSendingPermitted` could not return false in the middle of an iteration.

`PacketSendingPermitted` checks three pieces of state: 
- is slash record found? Note if a slash record is created or updated, the [iteration breaks anyways](https://github.com/cosmos/interchain-security/blob/18a86c359d219520e59bf6c1492780bfa212d8f6/x/ccv/consumer/keeper/relay.go#L222)
- is record.WaitingOnReply true? This flag is mutated from `OnAcknowledgementPacket`, which can't run in the middle of `SendPackets`. 
- Has retry delay period passed? `ctx.Blocktime()` will trivially not change in the middle of `SendPackets`

Regardless, it's possible I've missed something in my reasoning here. To protect against that possibility, we can simply `break` instead of `return` if `PacketSendingPermitted` returns false, ensuring the deletion line is always hit. I don't see any issues with that idea

Fixed in https://github.com/cosmos/interchain-security/commit/f57aa641139aa56b337bd64e516d841146e725f1